raster/r.carve/enforce_ds.c

```diff
@@ -76,32 +111,127 @@ int enforce_downstream(int infd, int outfd, struct Map_info *Map,
 
     G_message(_("Processing lines... "));
 
-    nlines = Vect_get_num_lines(Map);
-    for (line = 1; line <= nlines; line++)
-        retval = process_line(Map, outMap, rbuf, line, parm);
+    box_list = Vect_new_boxlist(0);
+
+    box.N = wind.north;
+    box.E = wind.east;
+    box.S = wind.south;
+    box.W = wind.west;
+    box.B = wind.bottom;
+    box.T = wind.top;
+
+    Vect_select_lines_by_box(Map, &box, GV_LINE, box_list);
+    field = Vect_get_field_number(Map, parm->field->answer);
+
+    /* Get key col */
+    finfo = Vect_get_field(Map, field);
+
+    if (parm->width_col->answer || parm->depth_col->answer) {
+        sql = create_select_sql_statement(Map, Fi, box_list, columns, &field,
+                                          finfo->key);
+
+        if (db_open_select_cursor(driver, sql.sql, &cursor, DB_SEQUENTIAL) !=
+            DB_OK)
+            G_fatal_error(_("Unable to open select cursor"));
+
+        def_width = G_malloc(sizeof(double));
+        def_depth = G_malloc(sizeof(double));
+        memcpy(def_width, &parm->swidth, sizeof(double));
+        memcpy(def_depth, &parm->sdepth, sizeof(double));
+
+        table = db_get_cursor_table(&cursor);
+        G_debug(1, "Default width: %.2f, depth: %.2f", *def_width / distance,
+                *def_depth);
+
+        cat_col = db_get_table_column_by_name(table, finfo->key);
+        if (parm->width_col->answer) {
+            width_col =
+                db_get_table_column_by_name(table, columns[*width_col_pos]);
+            width_col_type = db_get_column_sqltype(width_col);
+        }
+        if (parm->depth_col->answer) {
+            depth_col =
+                db_get_table_column_by_name(table, columns[*depth_col_pos]);
+            depth_col_type = db_get_column_sqltype(depth_col);
+        }
+
+        while (1) {
+            if (db_fetch(&cursor, DB_NEXT, &more) != DB_OK)
+                G_fatal_error(_("Unable to fetch data from table <%s>"),
+                              Fi->table);
+
+            if (!more)
+                break;
+
+            int cat = db_get_value_int(db_get_column_value(cat_col));
+
+            /* Set width */
+            set_value(width_col, &width_col_type, parm->width_col->answer,
+                      &parm->swidth, def_width, table, &wind, WIDTH);
+            parm->swidth /= distance;
+            /* Set depth */
+            set_value(depth_col, &depth_col_type, parm->depth_col->answer,
+                      &parm->sdepth, def_depth, table, &wind, DEPTH);
+
+            for (c = 0; c < sql.ncats; c++) {
```

Comment:
”comparison of integers of different signs: 'unsigned int' and 'int' `[-Werror,-Wsign-compare]” compiler warning